File: src/spectrl/codecs/numpress.py
"""MS-Numpress + zlib codecs.

Wraps the three MS-Numpress transforms (linear, slof, pic) with zlib. The raw
transforms come from one of two interchangeable, byte-identical backends:

- **pynumpress**: the C extension; preferred when importable.
- **pure Python** (:mod:`spectrl.codecs._numpress_py`): a dependency-free
  fallback used when ``pynumpress`` is not installed. This is what lets spectrl
  run in Pyodide / the browser (and anywhere a C toolchain is unavailable),
  where ``pynumpress`` has no wheel.

Backend selection is lazy: ``import spectrl`` never imports ``pynumpress``, so a
missing C extension is only a problem if you actually encode/decode a lossy
token, and even then the pure-Python path takes over transparently. Set
``SPECTRL_NUMPRESS_BACKEND=python`` (or ``pynumpress``) to pin one explicitly,
mainly for tests that assert the two backends agree.
"""

from __future__ import annotations

import math
import os
import struct
import zlib
from typing import Protocol

import numpy as np

from .._format import DEFAULT_NUMLIN_FP, DEFAULT_NUMSLOF_FP
from ._zlibutil import bounded_decompress

# Fixed points are whole numbers: the two defaults are, and a clamped slof fp is
# floored to one. Rounding a clamped fp DOWN is the safe direction -- the clamp
# exists to keep log(max + 1) * fp under the uint16 ceiling, and a smaller fp
# only pushes that product further below it -- and it lets the descriptor carry
# fp as a CBOR integer (3 or 5 bytes) instead of a float64 (9).
# SLOF fp must satisfy: log(max_intensity + 1) * fp <= 65535 (uint16 max)
# Use 3600 which handles intensities up to ~8e7; clip to a safe value if larger.
_SLOF_UINT16_MAX = 65535.0


def _safe_slof_fp(data: np.ndarray, desired_fp: float) -> int:
    """Return a slof fp that won't overflow uint16 given the array's max value.

    Always a whole number, floored so the clamp can only get safer.
    """
    max_val = float(np.max(data)) if len(data) > 0 else 1.0
    max_val = max(max_val, 1.0)
    max_fp = _SLOF_UINT16_MAX / (math.log(max_val + 1) + 1e-9)
    return max(1, int(min(desired_fp, max_fp)))


# ---------------------------------------------------------------------------
# Backend selection (lazy: never imports pynumpress at module import time)
# ---------------------------------------------------------------------------


class _NumpressBackend(Protocol):
    name: str

    def encode_linear(self, data: np.ndarray, fp: float) -> np.ndarray: ...
    def decode_linear(self, raw: np.ndarray) -> np.ndarray: ...
    def encode_slof(self, data: np.ndarray, fp: float) -> np.ndarray: ...
    def decode_slof(self, raw: np.ndarray) -> np.ndarray: ...
    def encode_pic(self, data: np.ndarray) -> np.ndarray: ...
    def decode_pic(self, raw: np.ndarray) -> np.ndarray: ...


def _validate_nibbles(raw: np.ndarray, offset: int) -> None:
    """Check packed integer framing before passing untrusted bytes to native code."""
    cursor = offset * 2
    end = len(raw) * 2
    while cursor < end:
        value = int(raw[cursor // 2])
        head = value & 15 if cursor % 2 else value >> 4
        if cursor == end - 1 and head == 0:
            return
        cursor += 1 + (8 - head if head <= 8 else 16 - head)
        if cursor > end:
            raise ValueError("truncated Numpress integer")


class _PynumpressBackend:
    name = "pynumpress"

    def __init__(self, mod) -> None:
        self._m = mod

    def encode_linear(self, data: np.ndarray, fp: float) -> np.ndarray:
        return np.asarray(self._m.encode_linear(data.astype(np.float64), fp), dtype=np.uint8)

    def decode_linear(self, raw: np.ndarray) -> np.ndarray:
        n = len(raw)
        # pynumpress 0.0.9 cannot decode a single-value linear blob (12 bytes: 8-byte
        # fixed point + one 4-byte int), though encode_linear emits exactly that. The
        # MS-Numpress reference decodes it (dataSize == 12 → one value); do so directly
        # to keep single-peak spectra round-trippable and cross-impl compatible.
        if n not in (8, 12) and n < 16:
            raise ValueError("truncated Numpress linear stream")
        if n >= 16:
            _validate_nibbles(raw, 16)
        if n == 12:
            buf = raw.tobytes()
            fixed_point = struct.unpack(">d", buf[:8])[0]
            first = int.from_bytes(buf[8:12], "little", signed=False)
            return np.array([first / fixed_point], dtype=np.float64)
        return np.array(self._m.decode_linear(raw), dtype=np.float64)

    def encode_slof(self, data: np.ndarray, fp: float) -> np.ndarray:
        return np.asarray(self._m.encode_slof(data.astype(np.float64), fp), dtype=np.uint8)

    def decode_slof(self, raw: np.ndarray) -> np.ndarray:
        if len(raw) < 8 or len(raw) % 2:
            raise ValueError("truncated Numpress slof stream")
        return np.array(self._m.decode_slof(raw), dtype=np.float64)

    def encode_pic(self, data: np.ndarray) -> np.ndarray:
        return np.asarray(self._m.encode_pic(data.astype(np.float64)), dtype=np.uint8)

    def decode_pic(self, raw: np.ndarray) -> np.ndarray:
        _validate_nibbles(raw, 0)
        return np.array(self._m.decode_pic(raw), dtype=np.float64)


class _PurePythonBackend:
    name = "python"

    def __init__(self) -> None:
        from . import _numpress_py as m

        self._m = m

    def encode_linear(self, data: np.ndarray, fp: float) -> np.ndarray:
        return self._m.encode_linear(data.astype(np.float64), fp)

    def decode_linear(self, raw: np.ndarray) -> np.ndarray:
        return self._m.decode_linear(raw)

    def encode_slof(self, data: np.ndarray, fp: float) -> np.ndarray:
        return self._m.encode_slof(data.astype(np.float64), fp)

    def decode_slof(self, raw: np.ndarray) -> np.ndarray:
        return self._m.decode_slof(raw)

    def encode_pic(self, data: np.ndarray) -> np.ndarray:
        return self._m.encode_pic(data.astype(np.float64))

    def decode_pic(self, raw: np.ndarray) -> np.ndarray:
        return self._m.decode_pic(raw)


_BACKEND: _NumpressBackend | None = None


def _resolve_backend() -> _NumpressBackend:
    global _BACKEND
    if _BACKEND is not None:
        return _BACKEND

    choice = os.environ.get("SPECTRL_NUMPRESS_BACKEND", "auto").strip().lower()
    if choice in ("python", "pure", "purepython"):
        _BACKEND = _PurePythonBackend()
    elif choice in ("pynumpress", "c"):
        import pynumpress  # explicit request: let ImportError surface

        _BACKEND = _PynumpressBackend(pynumpress)
    else:  # "auto": prefer the C extension, fall back to pure Python
        try:
            import pynumpress

            _BACKEND = _PynumpressBackend(pynumpress)
        except ImportError:
            _BACKEND = _PurePythonBackend()
    return _BACKEND


def active_backend() -> str:
    """Return the name of the resolved numpress backend (``"pynumpress"`` or ``"python"``)."""
    return _resolve_backend().name


# ---------------------------------------------------------------------------
# Public codecs (raw transform via the selected backend, then zlib)
# ---------------------------------------------------------------------------


def _reject_negatives(data: np.ndarray, codec: str) -> np.ndarray:
    """Raise a catchable ValueError on negative input.

    The slof transform computes log(v + 1), which breaks for negatives, and the
    linear transform's rounding differs between the C and pure-Python backends
    for negatives (truncation vs floor), so byte-identity would be lost.
    Callers that may hold negative values fall back to a lossless codec.
    """
    arr = np.asarray(data, dtype=np.float64)
    if arr.size and float(arr.min()) < 0:
        raise ValueError(
            f"MS-Numpress {codec} codec cannot encode negative values; "
            "use a lossless codec for arrays that may contain negatives."
        )
    return arr


def validate_linear_domain(data: np.ndarray, fp: float) -> None:
    """Reject values that cannot be represented by every Numpress backend."""
    arr = np.asarray(data, dtype=np.float64)
    if not math.isfinite(fp) or fp <= 0:
        raise ValueError("Numpress linear fixed point must be finite and positive")
    with np.errstate(over="ignore", invalid="ignore"):
        scaled = np.floor(arr * fp + 0.5)
    if not np.isfinite(scaled).all() or np.any(scaled < 0) or np.any(scaled > (2**53 - 1) // 4):
        raise ValueError("Numpress linear scaled values exceed the safe numeric range")
    if np.any(scaled[:2] > 0xFFFFFFFF):
        raise ValueError("Numpress linear initial values exceed uint32, use a lossless codec or a smaller fixed point")
    residual = scaled[2:] - 2 * scaled[1:-1] + scaled[:-2]
    if np.any(residual < -(2**31)) or np.any(residual > 2**31 - 1):
        raise ValueError("Numpress linear prediction residual exceeds int32, use a lossless codec")


def validate_pic_domain(data: np.ndarray) -> None:
    """PIC represents unsigned 32-bit whole numbers only."""
    arr = np.asarray(data, dtype=np.float64)
    if not np.isfinite(arr).all() or np.any(arr < 0) or np.any(arr > 0xFFFFFFFF) or np.any(arr != np.floor(arr)):
        raise ValueError("Numpress PIC requires whole numbers in the uint32 range")


def encode_numlin_raw(data: np.ndarray, fp: float | None = None) -> bytes:
    """Encode an array with the raw MS-Numpress linear transform."""
    arr = _reject_negatives(data, "linear")
    fp = fp if fp is not None else DEFAULT_NUMLIN_FP
    validate_linear_domain(arr, fp)
    encoded = _resolve_backend().encode_linear(arr, fp)
    return encoded.tobytes()


def decode_numlin_raw(blob: bytes) -> np.ndarray:
    """Decode a raw MS-Numpress linear stream."""
    return _resolve_backend().decode_linear(np.frombuffer(blob, dtype=np.uint8))


def encode_numlin_zlib(data: np.ndarray, fp: float | None = None) -> bytes:
    """Encode array with MS-Numpress linear prediction then zlib."""
    return zlib.compress(encode_numlin_raw(data, fp))


def decode_numlin_zlib(blob: bytes, max_bytes: int | None = None) -> np.ndarray:
    """Decode MS-Numpress linear + zlib blob back to float64 array."""
    decompressed = bounded_decompress(blob, max_bytes)
    return decode_numlin_raw(decompressed)


def encode_numslof_raw(data: np.ndarray, fp: float | None = None) -> bytes:
    """Encode an array with the raw MS-Numpress short logged float transform."""
    arr = _reject_negatives(data, "slof")
    desired = fp if fp is not None else DEFAULT_NUMSLOF_FP
    safe_fp = _safe_slof_fp(arr, desired)
    encoded = _resolve_backend().encode_slof(arr, safe_fp)
    return encoded.tobytes()


def decode_numslof_raw(blob: bytes) -> np.ndarray:
    """Decode a raw MS-Numpress short logged float stream."""
    return _resolve_backend().decode_slof(np.frombuffer(blob, dtype=np.uint8))


def encode_numslof_zlib(data: np.ndarray, fp: float | None = None) -> bytes:
    """Encode array with MS-Numpress short logged float then zlib."""
    return zlib.compress(encode_numslof_raw(data, fp))


def decode_numslof_zlib(blob: bytes, max_bytes: int | None = None) -> np.ndarray:
    """Decode MS-Numpress slof + zlib blob back to float64 array."""
    decompressed = bounded_decompress(blob, max_bytes)
    return decode_numslof_raw(decompressed)


def encode_numpic_raw(data: np.ndarray) -> bytes:
    """Encode an array with the raw MS-Numpress positive integer transform.

    MS-Numpress PIC represents only non-negative integers. pynumpress aborts the
    whole process (an uncatchable native C++ throw) on negative input, so reject
    negatives here with a catchable ``ValueError`` instead; callers that may hold
    negative values (e.g. a charge array with sentinels) should fall back to a
    lossless codec.
    """
    arr = _reject_negatives(data, "PIC")
    validate_pic_domain(arr)
    encoded = _resolve_backend().encode_pic(arr)
    return encoded.tobytes()


def decode_numpic_raw(blob: bytes) -> np.ndarray:
    """Decode a raw MS-Numpress positive integer stream."""
    return _resolve_backend().decode_pic(np.frombuffer(blob, dtype=np.uint8))


def encode_numpic_zlib(data: np.ndarray, fp: float | None = None) -> bytes:
    """Encode array with MS-Numpress positive integer then zlib."""
    return zlib.compress(encode_numpic_raw(data))


def decode_numpic_zlib(blob: bytes, max_bytes: int | None = None) -> np.ndarray:
    """Decode MS-Numpress pic + zlib blob back to float64 array."""
    decompressed = bounded_decompress(blob, max_bytes)
    return decode_numpic_raw(decompressed)

File: js/src/numpress.ts
/**
 * MS-Numpress codecs (linear, slof, pic), ported faithfully from the reference
 * C++ implementation (Teleman et al., ms-numpress). Byte-compatible with
 * pynumpress / the Python reference impl, so tokens round-trip across languages.
 *
 * Notes on the port:
 * - The fixed point is stored big-endian in the first 8 bytes (linear, slof).
 * - encodeInt/decodeInt operate on UNSIGNED 32-bit integers via `>>> 0`.
 * - 32-bit nibble packing matches the reference exactly.
 */

const FLOOR = Math.floor;

/**
 * Throw a catchable error on negative input. slof computes log(v + 1), which
 * breaks for negatives; linear and pic rounding of negatives differs across
 * implementations (truncation vs floor / unsigned wrap), so byte-identity
 * would be lost. Callers fall back to a lossless codec for such arrays.
 */
function rejectNegatives(data: Float64Array, codec: string): void {
  for (const v of data) {
    if (v < 0) {
      throw new Error(
        `MS-Numpress ${codec} codec cannot encode negative values; use a lossless codec for arrays that may contain negatives.`,
      );
    }
  }
}

// ── fixed point (big-endian float64) ──────────────────────────────────────────

function encodeFixedPoint(fixedPoint: number, out: Uint8Array, offset: number): void {
  const dv = new DataView(out.buffer, out.byteOffset + offset, 8);
  dv.setFloat64(0, fixedPoint, false); // big-endian
}

function decodeFixedPoint(data: Uint8Array): number {
  const dv = new DataView(data.buffer, data.byteOffset, 8);
  return dv.getFloat64(0, false);
}

// ── variable-length signed int as half-bytes (nibbles) ───────────────────────

/** Append the nibble encoding of unsigned-32 `x` into `half` at `offset`; return nibbles written. */
function encodeInt(x: number, half: Uint8Array, offset: number): number {
  x = x >>> 0;
  const mask = 0xf0000000;
  const init = (x & mask) >>> 0;
  let l: number;

  if (init === 0) {
    l = 8;
    for (let i = 0; i < 8; i++) {
      const m = (mask >>> (4 * i)) >>> 0;
      if (((x & m) >>> 0) !== 0) {
        l = i;
        break;
      }
    }
    half[offset] = l;
    for (let i = l; i < 8; i++) half[offset + 1 + i - l] = (x >>> (4 * (i - l))) & 0xf;
    return 1 + 8 - l;
  } else if (init === (mask >>> 0)) {
    l = 7;
    for (let i = 0; i < 8; i++) {
      const m = (mask >>> (4 * i)) >>> 0;
      if (((x & m) >>> 0) !== m) {
        l = i;
        break;
      }
    }
    half[offset] = l + 8;
    for (let i = l; i < 8; i++) half[offset + 1 + i - l] = (x >>> (4 * (i - l))) & 0xf;
    return 1 + 8 - l;
  } else {
    half[offset] = 0;
    for (let i = 0; i < 8; i++) half[offset + 1 + i] = (x >>> (4 * i)) & 0xf;
    return 9;
  }
}

interface IntState {
  di: number;
  half: number;
}

/** Decode one nibble-packed unsigned-32 int from `data`, advancing `st`.
 * Throws on truncated input rather than reading past the end. */
function decodeInt(data: Uint8Array, st: IntState): number {
  if (st.di >= data.length) throw new Error("numpress: truncated input");
  let head: number;
  if (st.half === 0) {
    head = data[st.di]! >> 4;
  } else {
    head = data[st.di]! & 0xf;
    st.di++;
  }
  st.half = 1 - st.half;

  let res = 0;
  let n: number;
  if (head <= 8) {
    n = head;
  } else {
    n = head - 8;
    const mask = 0xf0000000;
    for (let i = 0; i < n; i++) res = (res | ((mask >>> (4 * i)) >>> 0)) >>> 0;
  }

  if (n === 8) return res >>> 0;

  for (let i = n; i < 8; i++) {
    if (st.di >= data.length) throw new Error("numpress: truncated input");
    let hb: number;
    if (st.half === 0) {
      hb = data[st.di]! >> 4;
    } else {
      hb = data[st.di]! & 0xf;
      st.di++;
    }
    res = (res | (hb << ((i - n) * 4))) >>> 0;
    st.half = 1 - st.half;
  }
  return res >>> 0;
}

// little-endian 32-bit write using arithmetic (safe for values up to 2^32-1)
function writeU32LE(out: number[], v: number): void {
  v = v >>> 0 === v ? v : ((v % 0x100000000) + 0x100000000) % 0x100000000;
  out.push(v & 0xff, (FLOOR(v / 256) & 0xff) >>> 0, (FLOOR(v / 65536) & 0xff) >>> 0, (FLOOR(v / 16777216) & 0xff) >>> 0);
}

function readU32LE(data: Uint8Array, off: number): number {
  return data[off]! + data[off + 1]! * 256 + data[off + 2]! * 65536 + data[off + 3]! * 16777216;
}

// ── linear ────────────────────────────────────────────────────────────────────

export function validateLinearDomain(data: ArrayLike<number>, fp: number): void {
  if (!Number.isFinite(fp) || fp <= 0) throw new Error("Numpress linear fixed point must be finite and positive")
  let previous = 0
  let before = 0
  for (const [i, item] of Array.from(data).entries()) {
    const value = Math.floor(item * fp + 0.5)
    if (!Number.isSafeInteger(value) || value < 0 || value > Math.floor(Number.MAX_SAFE_INTEGER / 4)) {
      throw new Error("Numpress linear scaled values exceed the safe numeric range")
    }
    if (i < 2 && value > 0xffffffff) throw new Error("Numpress linear initial values exceed uint32, use a lossless codec or a smaller fixed point")
    const residual = value - 2 * previous + before
    if (i >= 2 && (residual < -2147483648 || residual > 2147483647)) {
      throw new Error("Numpress linear prediction residual exceeds int32, use a lossless codec")
    }
    before = previous
    previous = value
  }
}

export function validatePicDomain(data: ArrayLike<number>): void {
  for (const value of Array.from(data)) {
    if (!Number.isInteger(value) || value < 0 || value > 0xffffffff) {
      throw new Error("Numpress PIC requires whole numbers in the uint32 range")
    }
  }
}

export function encodeLinear(data: Float64Array, fixedPoint: number): Uint8Array {
  rejectNegatives(data, "linear")
  validateLinearDomain(data, fixedPoint)
  const fp = new Uint8Array(8);
  encodeFixedPoint(fixedPoint, fp, 0);
  const out: number[] = Array.from(fp);

  const n = data.length;
  if (n === 0) return Uint8Array.from(out);

  let i1 = FLOOR(data[0]! * fixedPoint + 0.5);
  writeU32LE(out, i1);
  if (n === 1) return Uint8Array.from(out);

  let i2 = FLOOR(data[1]! * fixedPoint + 0.5);
  writeU32LE(out, i2);

  const halfBytes = new Uint8Array(16);
  let halfByteCount = 0;

  for (let i = 2; i < n; i++) {
    const i0 = i1;
    i1 = i2;
    i2 = FLOOR(data[i]! * fixedPoint + 0.5);
    const extrapol = i1 + (i1 - i0);
    const diff = (i2 - extrapol) | 0; // 32-bit signed
    halfByteCount += encodeInt(diff >>> 0, halfBytes, halfByteCount);

    for (let hbi = 1; hbi < halfByteCount; hbi += 2) {
      out.push(((halfBytes[hbi - 1]! << 4) | (halfBytes[hbi]! & 0xf)) & 0xff);
    }
    if (halfByteCount % 2 !== 0) {
      halfBytes[0] = halfBytes[halfByteCount - 1]!;
      halfByteCount = 1;
    } else {
      halfByteCount = 0;
    }
  }
  if (halfByteCount === 1) out.push((halfBytes[0]! << 4) & 0xff);

  return Uint8Array.from(out);
}

export function decodeLinear(data: Uint8Array): Float64Array {
  const dataSize = data.length;
  if (dataSize === 8) return new Float64Array(0);
  if (dataSize < 8) throw new Error("numpress linear: truncated (no fixed point)");

  const fixedPoint = decodeFixedPoint(data);
  if (dataSize < 12) throw new Error("numpress linear: truncated (no first value)");

  const result: number[] = [];
  let i1 = readU32LE(data, 8);
  result.push(i1 / fixedPoint);
  if (dataSize === 12) return Float64Array.from(result);
  if (dataSize < 16) throw new Error("numpress linear: truncated (no second value)");

  let i2 = readU32LE(data, 12);
  result.push(i2 / fixedPoint);

  const st: IntState = { di: 16, half: 0 };
  while (st.di < dataSize) {
    if (st.di === dataSize - 1 && st.half === 1 && (data[st.di]! & 0xf) === 0) break;
    const i0 = i1;
    i1 = i2;
    const diff = decodeInt(data, st) | 0; // signed
    const extrapol = i1 + (i1 - i0);
    const y = extrapol + diff;
    result.push(y / fixedPoint);
    i2 = y;
  }
  return Float64Array.from(result);
}

// ── slof (short logged float) ─────────────────────────────────────────────────

export function encodeSlof(data: Float64Array, fixedPoint: number): Uint8Array {
  rejectNegatives(data, "slof");
  const out = new Uint8Array(8 + data.length * 2);
  encodeFixedPoint(fixedPoint, out, 0);
  let ri = 8;
  for (let i = 0; i < data.length; i++) {
    const x = FLOOR(Math.log(data[i]! + 1) * fixedPoint + 0.5) & 0xffff;
    out[ri++] = x & 0xff;
    out[ri++] = (x >> 8) & 0xff;
  }
  return out;
}

export function decodeSlof(data: Uint8Array): Float64Array {
  if (data.length < 8 || data.length % 2 !== 0) throw new Error("numpress slof: truncated (no fixed point)")
  const fixedPoint = decodeFixedPoint(data);
  const result = new Float64Array((data.length - 8) >> 1);
  let ri = 0;
  for (let i = 8; i + 1 < data.length; i += 2) {
    const x = (data[i]! | (data[i + 1]! << 8)) & 0xffff;
    result[ri++] = Math.exp(x / fixedPoint) - 1;
  }
  return result;
}

// ── pic (positive integer) ────────────────────────────────────────────────────

export function encodePic(data: Float64Array): Uint8Array {
  rejectNegatives(data, "pic")
  validatePicDomain(data)
  const out: number[] = [];
  const halfBytes = new Uint8Array(16);
  let halfByteCount = 0;
  for (let i = 0; i < data.length; i++) {
    const x = (FLOOR(data[i]! + 0.5) >>> 0) as number;
    halfByteCount += encodeInt(x, halfBytes, halfByteCount);
    for (let hbi = 1; hbi < halfByteCount; hbi += 2) {
      out.push(((halfBytes[hbi - 1]! << 4) | (halfBytes[hbi]! & 0xf)) & 0xff);
    }
    if (halfByteCount % 2 !== 0) {
      halfBytes[0] = halfBytes[halfByteCount - 1]!;
      halfByteCount = 1;
    } else {
      halfByteCount = 0;
    }
  }
  if (halfByteCount === 1) out.push((halfBytes[0]! << 4) & 0xff);
  return Uint8Array.from(out);
}

export function decodePic(data: Uint8Array): Float64Array {
  const dataSize = data.length;
  const result: number[] = [];
  const st: IntState = { di: 0, half: 0 };
  while (st.di < dataSize) {
    if (st.di === dataSize - 1 && st.half === 1 && (data[st.di]! & 0xf) === 0) break;
    result.push(decodeInt(data, st));
  }
  return Float64Array.from(result);
}
